### PR TITLE
chore: add paysg validation

### PR DESCRIPTION
**Pull request by pregnantboy**
## Improvements
- Move setting of api key and base url to a shared before request callback
- Use urlPathParams to populate url paths (as requested by @ogp-weeloong)
- Verify credentials by making a get request to list all payments

## Fixes
- Testing of connection in the My Apps page not reflecting the correct test result

## Tests
- Fix tests
- Add tests for verify connection

## How to test this PR
- [x] Existing api calls should work
- [x] Adding an invalid connection should fail

### CODE REVIEW

**Review by ogp-weeloong — APPROVED** [review]
LGTM

Tested
* Creating new PaySG connection from step
* Reconnecting from connections page
* Creating new PaySG connection from connections page
* Bad API key or bad payment service id will cause connection addition to error out
* PaySG create payments still work

File: packages/backend/src/apps/paysg/auth/verify-credentials.ts [hunk 1]
```diff
@@ -11,6 +11,28 @@ export default async function verifyCredentials(
     throw new Error('Invalid PaySG API key')
   }
 
+  const paymentServiceId = authData.paymentServiceId as string
+
+  if (!paymentServiceId || typeof paymentServiceId !== 'string') {
+    throw new Error('Payment Service ID must be set')
+  }
+
```
Comment (ogp-weeloong) [inline packages/backend/src/apps/paysg/auth/verify-credentials.ts:19]: Nit: since we're not doing paysg URL parsing yet, we can add this as a stopgap.

Left the 52 magic number inside code block for brevity; can move to a const or something.

```suggestion

if (
    !paymentServiceId.startsWith('payment_service_') ||
    paymentServiceId.length != 52 // prefix + UUID
  ) {
    throw new Error('Invalid Payment Service ID')
  } 

```

Comment (ogp-weeloong) [reply packages/backend/src/apps/paysg/auth/verify-credentials.ts:19]: Actually we can park this inside `addAuthHeader` to be more general - but if we do this we lose error messages, since errors get converted to `Unable to validate payment service id and api key` later down.

I'm thinking for UX, it's ok to duplicate the validation checks in both `addAuthHeader` and this `verifyCredentials` mutation. That or we remove the re-throw later down (see my comment later)

File: packages/backend/src/apps/paysg/auth/verify-credentials.ts [hunk 2]
```diff
@@ -11,6 +11,28 @@ export default async function verifyCredentials(
...
+        urlPathParams: {
+          paymentServiceId,
+        },
+      },
+    )
+  } catch (e) {
+    if (e.response?.status === 401 || e.response?.status === 403) {
+      throw new Error('Invalid credentials')
+    }
+    throw new Error('Unable to validate payment service id and api key')
```
Comment (ogp-weeloong) [inline packages/backend/src/apps/paysg/auth/verify-credentials.ts:33]: If we don't want to duplicate our validation code (see my [above](https://github.com/opengovsg/plumber/pull/474/files#r1527027656) comment)...

We could also do this instead so that we don't swallow potentially informative error messages (while still reducing risk of secrets from http requests / responses being leaked through errors).

```suggestion
    if (e instanceof HttpError) {
      if (e.response?.status === 401 || e.response?.status === 403) {
        throw new Error('Invalid credentials')
      } else {
        throw new Error('Error with PaySG HTTP request')
    }
    
    // Rethrow only the error message to minimize risk of secret leakage
    // But if secrets are stored in the error message then we're dead in the water.
    throw new Error(e.message)
```

But thinking double validation might be better because there is some non-trivial chance that in the future, one of us might to do something lower in the stack:
```
throw new Error(`${JSON.stringify(e)}`)
```

File: packages/backend/src/apps/paysg/common/add-auth-header.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import logger from '@/helpers/logger'
+
+import { getApiBaseUrl } from './api'
+
+const addAuthHeader: TBeforeRequest = async ($, requestConfig) => {
+  const { apiKey, paymentServiceId } = $.auth.data
+
+  if (typeof apiKey !== 'string' || typeof paymentServiceId !== 'string') {
+    logger.error({
+      erorr: 'API key and payment service ID must be set',
```
Comment (m0nggh) [inline packages/backend/src/apps/paysg/common/add-auth-header.ts:12]: typo here, will make a change in a separate pr
